### ISSUE TITLE
How to automatically retake snapshot after rsync error 23 using user-callback

### ISSUE THREAD

**Issue by jean-christophe-manciot** [bit-team/backintime#1573]
Here is the workflow inside my user-callback:
```
#!/bin/bash

profile_id="$1"
profile_name="$2"
backintime_status="$3"
...
if [[ ${profile_id} -eq 1 ]]; then
        # Working around a bug
        snapshot_log_file=/root/.local/share/backintime/takesnapshot_.log
else
        snapshot_log_file="/root/.local/share/backintime/takesnapshot_${profile_id}.log"
fi
case $backintime_status in
        1) #Backup process begins
                printLog "1 - backup process begins"
                ;;
        2) #Backup process ends
                printLog "2 - backup process ends"

                # Waiting for the log to be complete
                sleep 1m
                rsync_return_code=$(grep -P "'rsync' ended with exit code " "${snapshot_log_file}" | sed -E "s|^.*'rsync' ended with exit code ([0-9]+).*$|\1|g")

                case $profile_name in
                        "Main profile")
                                case "$rsync_return_code" in
                                        0|24)
                                                /usr/bin/nice -n 19 /usr/bin/ionice -c2 -n7 /usr/bin/backintime --profile "$next_profile_name" backup
                                                ;;
                                        23)
                                                # Retaking the backup
                                                /usr/bin/nice -n 19 /usr/bin/ionice -c2 -n7 /usr/bin/backintime --profile "$profile_name" backup
                                                ;;
                                        *)
                                                # Saving snapshot log
                                                cp --force --preserve=all "${snapshot_log_file}" "${snapshot_log_file}.$(date --iso-8601=minutes)"
                                                exit 1
                                                ;;
                                esac
                                ;;
                ...
                esac
        ...
esac
```

Unfortunately, this does not work when **rsync error 23** happens; I get in **takesnapshot.log**:
```
...
notify-send Back In Time (user) : profile_name Error: rsync: [sender] read errors mapping "filename": No data available (61)
notify-send Back In Time (user) : profile_name 'rsync' ended with exit code 23: See 'man rsync' for more details
notify-send Back In Time (user) : profile_name Failed to take snapshot 2023-11-19 19:01:14.'
ERROR: [plugins/usercallbackplugin.py:94 UserCallbackPlugin.callback] user-callback returned 'INFO: Lock
INFO: Take a new snapshot. Profile: profile_id profile_name
INFO: Call rsync to take the snapshot
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
WARNING: Command "rsync ..." returns 23
ERROR: Failed to take snapshot.

Back In Time
Version: 1.4.1

Back In Time comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type `backintime --license' for details.'
ERROR: user-callback returned 'WARNING: A backup is already running.  The pid of the already running backup is in file /root/.local/share/backintime/worker7.lock.  Maybe delete it'
WARNING: user-callback returncode: 1
ERROR: Plugin usercallbackplugin processEnd failed: 
INFO: Unlock'
WARNING: [plugins/usercallbackplugin.py:96 UserCallbackPlugin.callback] user-callback returncode: 1
ERROR: [common/pluginmanager.py:289 PluginManager.processEnd] Plugin usercallbackplugin processEnd failed: 
DEBUG: [common/tools.py:1103 keyringSupported] No keyring due to import error.
DEBUG: [common/mount.py:161 Mount.__init__] pw-cache is not running
DEBUG: [common/mount.py:169 Mount.__init__] Call command: /usr/bin/backintime pw-cache start
DEBUG: [plugins/usercallbackplugin.py:80 UserCallbackPlugin.callback] Call user-callback: /root/.config/backintime/user-callback 10 another_profile_name 8
INFO: [common/snapshots.py:847 Snapshots.backup] Unlock
```

Basically, it says the worker is locked and suggests to delete the lock file.
Is it safe to delete that file when the worker is still active before retaking the snapshot?

**bit-diags**:
```
backintime --diagnostics
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
{
    "backintime": {
        "name": "Back In Time",
        "version": "1.4.1",
        "latest-config-version": 6,
        "local-config-file": "/root/.config/backintime/config",
        "local-config-file-found": true,
        "global-config-file": "/etc/backintime/config",
        "global-config-file-found": false,
        "started-from": "/usr/share/backintime/common",
        "running-as-root": true,
        "user-callback": "/root/.config/backintime/user-callback",
        "keyring-supported": false
    },
    "host-setup": {
        "platform": "Linux-6.5.0-10-generic-x86_64-with-glibc2.38",
        "system": "Linux #10-Ubuntu SMP PREEMPT_DYNAMIC Fri Oct 13 13:49:38 UTC 2023",
        "OS": {
            "/etc/os-release": "Ubuntu 23.10",
            "/etc/lsb-release": "DISTRIB_ID=Ubuntu\nDISTRIB_RELEASE=23.10\nDISTRIB_CODENAME=mantic\nDISTRIB_DESCRIPTION=\"Ubuntu 23.10\"\n",
            "/etc/debian_version": "trixie/sid\n"
        },
        "display-system": "($XDG_SESSION_TYPE not set)",
        "locale": "en_US, UTF-8",
        "PATH": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin",
        "RSYNC_OLD_ARGS": "(not set)",
        "RSYNC_PROTECT_ARGS": "(not set)"
    },
    "python-setup": {
        "python": "3.11.6 main Oct  8 2023 05:06:43 CPython GCC 13.2.0",
        "python-executable": "/usr/bin/python3",
        "python-executable-symlink": true,
        "python-executable-resolved": "/usr/bin/python3.11",
        "sys.path": [
            "/usr/share/backintime/qt/plugins",
            "/usr/share/backintime/common/plugins",
            "/usr/share/backintime/plugins",
            "/usr/share/backintime/common",
            "/usr/lib/python311.zip",
            "/usr/lib/python3.11",
            "/usr/lib/python3.11/lib-dynload",
            "/usr/local/lib/python3.11/dist-packages",
            "/usr/lib/python3/dist-packages",
            "/usr/lib/python3.11/dist-packages"
        ],
        "qt": {
            "Version": "PyQt 5.15.9 / Qt 5.15.10",
            "Theme": "hicolor",
            "Theme Search Paths": [
                "/root/.local/share/icons",
                "/usr/local/share/icons",
                "/usr/share/icons",
                ":/icons"
            ],
            "Fallback Theme": "hicolor",
            "Fallback Search Paths": [
                "/usr/share/pixmaps"
            ]
        }
    },
    "external-programs": {
        "rsync": {
            "version": "3.2.7",
            "protocol": "31.0",
            "capabilities": {
                "file_bits": 64,
                "inum_bits": 64,
                "timestamp_bits": 64,
                "long_int_bits": 64,
                "socketpairs": true,
                "symlinks": true,
                "symtimes": true,
                "hardlinks": true,
                "hardlink_specials": true,
                "hardlink_symlinks": true,
                "IPv6": true,
                "atimes": true,
                "batchfiles": true,
                "inplace": true,
                "append": true,
                "ACLs": true,
                "xattrs": true,
                "secluded_args": "optional",
                "iconv": true,
                "prealloc": true,
                "stop_at": true,
                "crtimes": false
            },
            "optimizations": {
                "SIMD_roll": true,
                "asm_roll": false,
                "openssl_crypto": true,
                "asm_MD5": false
            },
            "checksum_list": [
                "xxh128",
                "xxh3",
                "xxh64",
                "md5",
                "md4",
                "sha1",
                "none"
            ],
            "compress_list": [
                "zstd",
                "lz4",
                "zlibx",
                "zlib",
                "none"
            ],
            "daemon_auth_list": [
                "sha512",
                "sha256",
                "sha1",
                "md5",
                "md4"
            ]
        },
        "ssh": "OpenSSH_9.3p1 Ubuntu-1ubuntu3, OpenSSL 3.0.10 1 Aug 2023",
        "sshfs": "3.7.3",
        "encfs": "(no encfs)",
        "shell": "/bin/bash",
        "shell-version": "GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu)"
    }
}
```

**Comment by aryoda** [comment]
```
ERROR: user-callback returned 'WARNING: A backup is already running.  The pid of the already running backup is in file /root/.local/share/backintime/worker7.lock.  Maybe delete it'
```

> Is it safe to delete that file when the worker is still active before retaking the snapshot?

I think it is not safe to delete the lock file of the same backup profile and start another BiT instance for the same backup profile while BiT is still running (even though `rsync` has already finished).

The reason is that the snapshot lock file contains the PID (AFAIR) and once the first instance exits it would premeaturely delete the lock file of the second started BiT causing side effects (PID no longer available, another instance could restart etc.).

I will look into the code the next days to get an exact and clear picture when and how the lock file is used and what happens if it is deleted prematurely. Perhaps (but very unlikely) it would be possible to make BiT delete the lock file immediately after `rsync` finished but **before** doing the next user-callback call (requires all internal BiT data to be written and still the global BiT app instance lock may be a show stopper then).

IMHO it would be better if we add a new "retry feature" to BiT since
- we now have the rsync exit codes that could be used to decide on a retry (even user configurable)
- BiT would retry to backup only the missing files if we restart `rsync` for the same (incomplete) snapshot.

**Comment by jean-christophe-manciot** [comment]
@aryoda 
Of course, a new 'retry' feature would be preferable.

**Comment by aryoda** [comment]
@jean-christophe-manciot

I am still trying to understand how our code exactly works to answer your question (it is more complicated than I thought).

Currently I am testing the impact of the used advisory lock on another "signal" file used to serialize the execution of multiple instances (more exactly: Does it serialize the execution despite the logged warning `WARNING: A backup is already running.`)?

https://github.com/bit-team/backintime/blob/59097f7a3c0ddd978e5a7a47b21909bed637a010/common/applicationinstance.py#L133-L148

@buhtz I am writing a doc-dev summary about the locking while "reverse-engineering" the logic

**Comment by aryoda** [comment]
@jean-christophe-manciot I have understood the BiT implementation of the lock files now good enough to answer your question and have good news:

> Is it safe to delete that file when the worker is still active before retaking the snapshot?

## TLDR ;-)

Yes, if you
- delete the `worker<Profile ID>.lock` in the user-callback handler for the event number 2 ("process end")
- and start a new `backintime` process for the same profile

the running backup ends gracefully because the deletion of the non-existing lock file ignores any error (here: missing lock file) and new backup process does also work.

## Caveats

There may be rare race conditions, eg.
- if for some reasons another BiT process is started for the same profile in parallel only one BiT process does really take a new snapshot (and the other stops with the well known message `WARNING: A backup is already running.`).
- if another BiT _restore_ process is running another lock file named `restore<Profile ID>.lock` may prevent taking a new snapshot and you should not delete it (because I have not yet checked how this works - but is on my list for the dev doc of lock files)

I think your implementation needs to consider at least these cases:

1. How to avoid endlessly starting a new BiT process because it always ends with an (`rsync`) error (like a "fork bomb" ;-)
2. Do you really want to start another time-consuming BiT process (disk memory is not wasted due to hard-links, but a new process means to create a completely new snapshot in a new snapshot folder) or could you possible try to grep and resume the `rsync` command for _the same snapshot_ - but this does not leave traces in the BiT snapshot logs).

## Possible alternative without deleting the lock file

- Start the new BiT process with a delay of a few seconds (eg. 30). This should the running BiT process give enough time to end normally 

## How I did test this

I have started a backup via `backintime --debug --profile-id 1 backup` in a Python debug session and stopped a breakpoint exactly where the user-callback event number 2 is called:

https://github.com/bit-team/backintime/blob/726bffa624d9a6da4ebd13ceb40721e1290e1eaf/common/snapshots.py#L830

I have then
1. started the same backup as a new process in a terminal (without deleting the `worker.lock` file) and it did exit  with `WARNING: A backup is already running.`
2. deleted the `worker.lock` file
3. started the same backup as a new process in a terminal  and it did succeed
4. continued the first backup process in the Python debug session and it did succeed

**Disclaimer: So there is a slight risk that starting another BiT process from within the user-callback bash script may behave differently!**

## Next steps

1. DONE: I will finish my dev doc for lock files (restore and other actions are still missing) and send a PR.
    See: https://github.com/bit-team/backintime/blob/dev/common/doc-dev/4_Control_files_usage_(locks_flocks_logs_and_others).md
3. DONE (#1591): I will create a new feature request issue to implement a "retry" feature within BiT (which works more efficiently because it "continues" the failed snapshot via `rsync --append ...` or similar instead of creating a new "full" snapshot)
4. DONE: I would close this issue (if it is OK for you) after sending and merging my dev doc PR for lock files

**Comment by jean-christophe-manciot** [comment]
@aryoda 
> delete the worker<Profile ID>.lock in the user-callback handler for the event number 2 ("process end")

Except for the 'Main profile' which uses```worker.lock```.

**Comment by aryoda** [comment]
I am closing this now since it is answered.

I have also added a new FR for a retry feature in BiT: #1591
